src/plonkish/backend/powdr_pil.rs

```diff
@@ -0,0 +1,418 @@
+    let super_asts = super_circuit.get_super_asts();
+    assert!(super_asts.len() == circuit_names.len());
+
+    // Get annotations map for the super circuit, which is a HashMap of object UUID to object
+    // annotation.
+    let mut super_circuit_annotations_map: HashMap<UUID, String> = HashMap::new();
+
+    // Loop over each AST.
+    for (ast, circuit_name) in super_asts.iter().zip(circuit_names.iter()) {
+        // Create `annotations_map` for each AST, to be added to `super_circuit_annotations_map`.
+        let mut annotations_map: HashMap<UUID, String> = HashMap::new();
+
+        // First, get AST level annotations.
+        annotations_map.extend(ast.annotations.clone());
+
+        // Second, get step level annotations.
+        for step_type in ast.step_types.values() {
+            annotations_map.extend(step_type.annotations.clone());
+        }
+
+        // Convert annotation to circuit_name.annotation, because this is the general format of
+        // referring to variables in PIL is there are more than one circuit.
+        super_circuit_annotations_map.extend(annotations_map.into_iter().map(
+            |(uuid, annotation)| {
+                (
+                    uuid,
+                    format!("{}.{}", circuit_name.clone(), annotation.replace(' ', "_")),
+                )
+            },
+        ));
+
+        // Finally, get annotations for the circuit names.
+        super_circuit_annotations_map.insert(ast.id, circuit_name.clone());
+    }
+
+    // Create a mapping from AST id to IR id, which is needed to link from AST to TraceWitness,
+    // because TraceWitness only stores IR id.
+    let ast_id_to_ir_id_mapping = super_circuit.get_ast_id_to_ir_id_mapping();
+
+    let mut pil = String::new(); // The string to return.
+
+    // For each AST, find its corresponding TraceWitness. Note that some AST might not have a
+    // corresponding TraceWitness, so witness is an Option.
+    for (ast, circuit_name) in super_asts.iter().zip(circuit_names.iter()) {
+        let witness = super_trace_witnesses.get(ast_id_to_ir_id_mapping.get(&ast.id).unwrap());
+
+        // Create PIL IR
+        let pil_ir =
+            PilCircuit::from_ast_and_witness(ast.clone(), witness.cloned(), circuit_name.clone());
+
+        // Generate PIL code given the IR and annotations map.
+        let pil_circuit = generate_pil_given_annotations(&pil_ir, &super_circuit_annotations_map);
+        writeln!(pil, "{}", pil_circuit).unwrap();
+    }
+
+    pil
+}
+
+// Generate PIL code given the IR of a single PIL circuit. Need to supply `annotations_map`,
+// which only need to contain annotations for the circuit's own components if we are converting
+// a standalone circuit. If we are converting a super circuit, `annotations_map` also needs to
+// contain annotations for components outside the circuit.
+fn generate_pil_given_annotations<F: Clone + Debug>(
```

Comment:
@qwang98 this function is too long, please split it

Reply:
Fixed.